File: ITDashboard.API/Models/DashboardModels.cs
using System.Text.Json.Serialization;

namespace ITDashboard.API.Models;

// ─── Entity (Organisation/Client) ───────────────────────────
public class EntityModel
{
    public int Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public string? Description { get; set; }
    public string DisplayName { get; set; } = string.Empty;
    public string? Address { get; set; }
    public string? Logo { get; set; }
    public string? ContactEmail { get; set; }
    public bool IsActive { get; set; }
    public DateTime CreatedDate { get; set; }
}

// ─── Master Configuration (dropdowns) ───────────────────────
public class MasterConfigModel
{
    public int Id { get; set; }
    public string FieldType { get; set; } = string.Empty;
    public string FieldName { get; set; } = string.Empty;
    public string? FieldValues { get; set; }
    public string? EntityIds { get; set; }
    public string? IsMandatory { get; set; }
    public string? IsActive { get; set; }
}

// ─── Category ────────────────────────────────────────────────
public class CategoryModel
{
    public int Id { get; set; }
    public string? EntityIds { get; set; }
    public string CategoryName { get; set; } = string.Empty;
    public string? CategoryDescription { get; set; }
    public string? IsActive { get; set; }
    public int? DepartmentId { get; set; }
    public string? DepartmentName { get; set; }
    public DateTime? CreatedDate { get; set; }
}

// ─── Subcategory ─────────────────────────────────────────────
public class SubcategoryModel
{
    public int Id { get; set; }
    public string? EntityIds { get; set; }
    public string SubcategoryName { get; set; } = string.Empty;
    public string? SubcategoryDescription { get; set; }
    public string? IsActive { get; set; }
    public int CategoryId { get; set; }
    public string? CategoryName { get; set; }
    public DateTime? CreatedDate { get; set; }
}

// ─── Ticket ──────────────────────────────────────────────────
public class TicketModel
{
    public bool? IsActive { get; set; } = true;
    public string? ReleaseNote { get; set; }
    public string? TypeName { get; set; }
    public string? DepartmentName { get; set; }
    public string? StatusName { get; set; }
    public string? ScheduleBuildNo { get; set; }
    public DateTime? ScheduleBuildDate { get; set; }
    public DateTime? PlannedDate { get; set; }

    public long TotalTickets { get; set; }
    public long FunctionalCount { get; set; }
    public long TechnicalCount { get; set; }
    public long TestingCount { get; set; }
    public long DeploymentCount { get; set; }
    public long InProduction { get; set; }

    public int Id { get; set; }
    public int TicketNo { get; set; }
    public string Title { get; set; } = string.Empty;
    public string? Description { get; set; }
    public string? Assignee { get; set; }
    public int? EntityId { get; set; }
    public string? EntityName { get; set; }
    public int? CategoryId { get; set; }
    public string? CategoryName { get; set; }
    public int? DepartmentId { get; set; }
    public int? PlatformId { get; set; }
    public string? PlatformName { get; set; }
    public int? PriorityId { get; set; }
    public string? PriorityName { get; set; }
    public int? StatusId { get; set; }
    public int? TypeId { get; set; }
    public string? Remarks { get; set; }
    public DateTime CreatedDate { get; set; }
    public DateTime? UpdatedDate { get; set; }

    public string? assigned_users { get; set; }

    // Or keep both
    [JsonIgnore]
    public List<int>? AssignedUsers { get; set; }
}

// ─── Dashboard KPI ───────────────────────────────────────────
public class DashboardKpiModel
{
    public long TotalTickets { get; set; }
    public long YetToStart { get; set; }
    public long OpenTickets { get; set; }
    public long InProgress { get; set; }
    public long ClosedTickets { get; set; }
    public long OverdueTickets { get; set; }
    public long TotalCr { get; set; }
    public long TotalIncidents { get; set; }
}

//// ─── Portfolio Summary ───────────────────────────────────────
//public class PortfolioSummaryModel
//{
//    public int? EntityId { get; set; }
//    public string? EntityName { get; set; }
//    public int? CategoryId { get; set; }
//    public string? CategoryName { get; set; }
//    public long TotalTickets { get; set; }
//    public long OpenTickets { get; set; }
//    public long InProduction { get; set; }
//    public long Overdue { get; set; }
//}

// ─── CR Pipeline ─────────────────────────────────────────────
public class CrPipelineModel
{
    public int Id { get; set; }
    public int TicketNo { get; set; }
    public string Title { get; set; } = string.Empty;
    public string? Assignee { get; set; }
    public string? EntityName { get; set; }
    public string? CategoryName { get; set; }
    public string? PriorityName { get; set; }
    public string? StatusName { get; set; }
    public string? ScheduleBuildNo { get; set; }
    public DateTime? PlannedDate { get; set; }
    public string? Remarks { get; set; }
    public DateTime CreatedDate { get; set; }
}

// ─── Release History ─────────────────────────────────────────
public class ReleaseHistoryModel
{
    public int Id { get; set; }
    public int TicketNo { get; set; }
    public string Title { get; set; } = string.Empty;
    public string? Assignee { get; set; }
    public string? EntityName { get; set; }
    public string? CategoryName { get; set; }
    public string? PlatformName { get; set; }
    public string? StatusName { get; set; }
    public string? ScheduleBuildNo { get; set; }
    public DateTime? ScheduleBuildDate { get; set; }
    public DateTime? PlannedDate { get; set; }
    public string? PlannedDuration { get; set; }
    public string? Remarks { get; set; }
}

// ─── Alert ───────────────────────────────────────────────────
public class AlertModel
{
    public string? AlertType { get; set; }
    public int TicketId { get; set; }
    public int TicketNo { get; set; }
    public string Title { get; set; } = string.Empty;
    public string? Assignee { get; set; }
    public string? EntityName { get; set; }
    public string? Priority { get; set; }
    public int DaysOverdue { get; set; }
    public string? StatusName { get; set; }
}

public class ReleaseDetailModel
{
    public string? ReleaseNo { get; set; }
    public DateTime? PlannedDate { get; set; }
    public DateTime? ActualReleaseDate { get; set; }
    public string? Status { get; set; }
}


// ─── Filter Request DTOs ──────────────────────────────────────
public class TicketFilterRequest
{
    public int? EntityId { get; set; }
    public int? StatusId { get; set; }
    public int? TypeId { get; set; }
    public int? PriorityId { get; set; }
    public DateTime? FromDate { get; set; }
    public DateTime? ToDate { get; set; }
}
// ITDashboard.API/Models/DashboardModels.cs

// Add this new model
public class PlatformModel
{
    public int Id { get; set; }
    public string FieldType { get; set; } = string.Empty;
    public string FieldName { get; set; } = string.Empty;
    public string? FieldValues { get; set; }
    public List<int> EntityIds { get; set; } = new List<int>();
    public string? IsMandatory { get; set; }
    public string? IsActive { get; set; }
}

// Update PortfolioSummaryModel to include platforms
public class PortfolioSummaryModel
{
    public int? EntityId { get; set; }
    public string? EntityName { get; set; }
    public int? CategoryId { get; set; }
    public string? CategoryName { get; set; }
    public long TotalTickets { get; set; }
    public long OpenTickets { get; set; }
    public long InProgress { get; set; }
    public long InProduction { get; set; }
    public long Overdue { get; set; }
    public List<PlatformInfo>? Platforms { get; set; }  // Add this
}

// Add this helper model
public class PlatformInfo
{
    public int Id { get; set; }
    public string Name { get; set; } = string.Empty;
}
// Key Risks Model
public class KeyRiskModel
{
    public string? Category { get; set; }
    public int Id { get; set; }
    public string Title { get; set; } = string.Empty;
    public string? Description { get; set; }
    public string Severity { get; set; } = "Medium";
    public string Status { get; set; } = "Open";
    public string? AssignedTo { get; set; }
    public DateTime? DueDate { get; set; }
    public string? MitigationPlan { get; set; }
    public int? EntityId { get; set; }
    public DateTime CreatedDate { get; set; }
    public DateTime UpdatedDate { get; set; }
    public bool IsActive { get; set; }
}

// Create/Update Request Models
public class CreateKeyRiskRequest
{
    public string Title { get; set; } = string.Empty;
    public string? Description { get; set; }
    public string Severity { get; set; } = "Medium";
    public string Status { get; set; } = "Open";
    public string? AssignedTo { get; set; }
    public DateTime? DueDate { get; set; }
    public string? MitigationPlan { get; set; }
    public int? EntityId { get; set; }
    public string? Category { get; set; }
}

public class UpdateKeyRiskRequest
{
    public string Title { get; set; } = string.Empty;
    public string? Description { get; set; }
    public string Severity { get; set; } = "Medium";
    public string Status { get; set; } = "Open";
    public string? AssignedTo { get; set; }
    public DateTime? DueDate { get; set; }
    public string? MitigationPlan { get; set; }
    public string? Category { get; set; }
}

public class UserModel
{
    public int Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public string? Realname { get; set; }
    public string? Email { get; set; }
}
public class BuildDocumentModel
{
    public int Id { get; set; }
    public string BuildNo { get; set; } = "";
    public int? PlatformId { get; set; }
    public int? EntityId { get; set; }
    public string FileName { get; set; } = "";
    public string FilePath { get; set; } = "";
    public long? FileSize { get; set; }
    public string? UploadedBy { get; set; }
    public DateTime? UploadedDate { get; set; }
}
// Models/CreateTicketRequest.cs
public class CreateTicketRequest
{
    public string Title { get; set; } = string.Empty;
    public string? Description { get; set; }
    public int EntityId { get; set; }
    public int PlatformId { get; set; }
    public int StatusId { get; set; }
    public int TypeId { get; set; }
    public int? PriorityId { get; set; }
    public int? DepartmentId { get; set; }
    public int? CategoryId { get; set; }
    public DateTime? PlannedDate { get; set; }
    public DateTime? ScheduleBuildDate { get; set; }
    public string? ScheduleBuildNo { get; set; }  // null = Future, value = Current
    public int? AssignedUserId { get; set; }
}

// Models/AuditLogModel.cs
public class AuditLogModel
{
    public int Id { get; set; }
    public int TicketId { get; set; }
    public string Action { get; set; } = string.Empty;
    public string? FieldName { get; set; }
    public string? OldValue { get; set; }
    public string? NewValue { get; set; }
    public string? ChangedBy { get; set; }
    public int? ChangedById { get; set; }
    public DateTime ChangedDate { get; set; }
    public string? Remarks { get; set; }
}

File: ITDashboard.UI/src/app/models/dashboard.models.ts
// src/app/models/dashboard.models.ts

export interface EntityModel {
    id: number;
    name: string;
    description?: string;
    displayName: string;
    address?: string;
    logo?: string;
    contactEmail?: string;
    isActive: boolean;
    createdDate: string;
}

export interface MasterConfigModel {
    id: number;
    fieldType: string;
    referrenceTo: string | null;
    fieldName: string;
    fieldValues: string;
    entityIds: string;
    isMandatory: string;
    isActive: string;
    createdDate: string;
    createdBy: string;
    updatedBy: string;
    updatedDate: string;
}

export interface CategoryModel {
    id: number;
    entityIds?: string;
    categoryName: string;
    categoryDescription?: string;
    isActive?: string;
    departmentId?: number;
    departmentName?: string;
    createdDate?: string;
}

export interface SubcategoryModel {
    id: number;
    subcategoryName: string;
    subcategoryDescription?: string;
    categoryId: number;
    categoryName?: string;
}

// Raw row returned by sp_get_department_summary
export interface DepartmentSummaryRow {
    id: number;
    ticketNo: number;
    title: string;
    assignee?: string;
    entityId?: number;
    entityName?: string;
    categoryName?: string;
    departmentName?: string;
    platformName?: string;
    priorityName?: string;
    statusName?: string;
    typeName?: string;
    scheduleBuildNo?: string;
    scheduleBuildDate?: string;
    plannedDate?: string;
    remarks?: string;
    createdDate: string;
    functionalCount: number;
    technicalCount: number;
    testingCount: number;
    deploymentCount: number;
    inProduction: number;
    totalTickets: number;
}

// Grouped shape used in the HTML template
export interface DeptGroupRow {
    departmentName: string;
    totalTickets: number;
    functionalCount: number;
    technicalCount: number;
    testingCount: number;
    deploymentCount: number;
    inProduction: number;
    releases: DepartmentSummaryRow[];
}

export interface TicketModel {
    id: number;
    ticketNo: number;
    title: string;
    description?: string;
    assignee?: string;
    entityId?: number;
    entityName?: string;
    categoryId?: number;
    categoryName?: string;
    departmentId?: number;
    departmentName?: string;
    platformId?: number;
    platformName?: string;
    priorityId?: number;
    priorityName?: string;
    statusId?: number;
    statusName?: string;
    typeId?: number;
    typeName?: string;
    scheduleBuildNo?: string;
    plannedDate?: string;
    scheduleBuildDate?: string;
    remarks?: string;
    createdDate: string;
    updatedDate?: string;
}

export interface DashboardKpiModel {
    totalTickets: number;
    yetToStart: number;
    openTickets: number;
    inProgress: number;
    closedTickets: number;
    overdueTickets: number;
    totalCr: number;
    totalIncidents: number;
      
}

export interface PortfolioSummaryModel {
    entityId?: number;
    entityName?: string;
    displayOrder?: number; 
    categoryId?: number;
    categoryName?: string;
    totalTickets: number;
    openTickets: number;
    inProduction: number;
    overdue: number;
    platforms?: PlatformInfo[];
    inProgress: number;
    yetToStart: number;

}

export interface CrPipelineModel {
    id: number;
    ticketNo: number;
    title: string;
    assignee?: string;
    entityName?: string;
    categoryName?: string;
    priorityName?: string;
    statusName?: string;
    scheduleBuildNo?: string;
    plannedDate?: string;
    remarks?: string;
    createdDate: string;
    platformId?: number;        // ← Add this
    platformName?: string;
}

export interface ReleaseHistoryModel {
    id: number;
    ticketNo: number;
    title: string;
    assignee?: string;
    entityName?: string;
    categoryName?: string;
    platformName?: string;
    statusName?: string;
    scheduleBuildNo?: string;
    scheduleBuildDate?: string;
    plannedDate?: string;
    plannedDuration?: string;
    remarks?: string;
}

export interface AlertModel {
    alertType?: string;
    ticketId: number;
    ticketNo: number;
    title: string;
    assignee?: string;
    entityName?: string;
    priority?: string;
    daysOverdue: number;
    statusName?: string;
}

export interface TicketFilterRequest {
    entityId?: number | null;
    statusId?: number | null;
    typeId?: number | null;
    priorityId?: number | null;
    fromDate?: string | null;
    toDate?: string | null;
}

// Platform related interfaces (defined only once)
export interface PlatformModel {
    id: number;
    fieldType: string;
    fieldName: string;
    fieldValues: string;
    entityIds: number[];
    isMandatory: string;
    isActive: string;
}

export interface PlatformInfo {
    id: number;
    name: string;
}
export interface ReleaseNote {
    id?: number;
    ticketId: number;
    entityId: number;
    platformId: number;
    buildNo: string;
    releaseNoteText: string;
    createdAt?: Date;
    updatedAt?: Date;
}
export interface KeyRiskModel {
    id: number;
    title: string;
    description: string;
    severity: 'Low' | 'Medium' | 'High' | 'Critical';
    status: 'Open' | 'In Progress' | 'Resolved' | 'Closed';
    assignedTo: string;
    dueDate: string;
    createdDate: string;
    updatedDate: string;
    mitigationPlan?: string;
}

// Also add these request models if needed
export interface CreateKeyRiskRequest {
    title: string;
    description?: string;
    severity: string;
    status: string;
    assignedTo?: string;
    dueDate?: string;
    mitigationPlan?: string;
    entityId?: number;
}

export interface UpdateKeyRiskRequest {
    title: string;
    description?: string;
    severity: string;
    status: string;
    assignedTo?: string;
    dueDate?: string;
    mitigationPlan?: string;
}
export interface CategoryModel {
    id: number;
    entityIds?: string;
    categoryName: string;
    categoryDescription?: string;
    isActive?: string;
    departmentId?: number;
    departmentName?: string;
    createdDate?: string;
}
export interface StatusModel {
    id: number;
    entityIds?: string;
    Status: string;
    Statusid: number;
    createdDate?: string;
    UpdatedDate?: String;
}